tests/utilTest.cpp

```diff
@@ -0,0 +1,421 @@
+
+    bytes = { 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff };
+    expected = 0xffffffffffffffffULL;
+    val = libzerocash::convertBytesVectorToInt(bytes);
+    BOOST_CHECK( expected == val );
+
+    bytes = { 0x80, 0x00, 0x00, 0x00, 0x80, 0x00, 0x00, 0x01 };
+    expected = 0x8000000080000001ULL;
+    val = libzerocash::convertBytesVectorToInt(bytes);
+    BOOST_CHECK( expected == val );
+
+    bytes = { 0xde, 0xad, 0xbe, 0xef }; // opposite of truncation
+    expected = 0xdeadbeefULL;
+    val = libzerocash::convertBytesVectorToInt(bytes);
+    BOOST_CHECK( expected == val );
+}
+
+BOOST_AUTO_TEST_CASE( testConcatenateTwoBoolVectors ) {
+    std::vector<bool> front = { 0, 1, 0 };
+    std::vector<bool> back = { 1, 0, 1 };
+    std::vector<bool> expected = { 0, 1, 0, 1, 0, 1 };
+    std::vector<bool> actual;
+    libzerocash::concatenateVectors(front, back, actual);
+    BOOST_CHECK( expected == actual );
+}
+
+BOOST_AUTO_TEST_CASE( testConcatenateTwoByteVectors ) {
+    std::vector<unsigned char> front = { 0, 1, 0 };
+    std::vector<unsigned char> back = { 1, 0, 1 };
+    std::vector<unsigned char> expected = { 0, 1, 0, 1, 0, 1 };
+    std::vector<unsigned char> actual;
+    libzerocash::concatenateVectors(front, back, actual);
+    BOOST_CHECK( expected == actual );
+}
+
+BOOST_AUTO_TEST_CASE( testConcatenateThreeBoolVectors ) {
+    std::vector<bool> front = { 0, 1, 0 };
+    std::vector<bool> middle { 1, 1, 1 };
+    std::vector<bool> back = { 1, 0, 1 };
+    std::vector<bool> expected = { 0, 1, 0, 1, 1, 1, 1, 0, 1 };
+    std::vector<bool> actual;
+    libzerocash::concatenateVectors(front, middle, back, actual);
+    BOOST_CHECK( expected == actual );
+}
+
+BOOST_AUTO_TEST_CASE( testConcatenateThreeByteVectors ) {
+    std::vector<unsigned char> front = { 0, 1, 0 };
+    std::vector<unsigned char> middle { 1, 1, 1 };
+    std::vector<unsigned char> back = { 1, 0, 1 };
+    std::vector<unsigned char> expected = { 0, 1, 0, 1, 1, 1, 1, 0, 1 };
+    std::vector<unsigned char> actual;
+    libzerocash::concatenateVectors(front, middle, back, actual);
+    BOOST_CHECK( expected == actual );
+}
+
+BOOST_AUTO_TEST_CASE( testSHA256TestVectors ) {
+    unsigned char actual_hash[32];
+    libzerocash::sha256(sha256_preimage, actual_hash, 3);
+    BOOST_CHECK( memcmp(sha256_hash, actual_hash, 32) );
+}
+
+BOOST_AUTO_TEST_CASE( testSHA256TestVectorsCTX ) {
+    unsigned char actual_hash[32];
+    SHA256_CTX_mod ctx256;
+    libzerocash::sha256(&ctx256, sha256_preimage, actual_hash, 3);
+    BOOST_CHECK( memcmp(sha256_hash, actual_hash, 32) );
+}
+
+BOOST_AUTO_TEST_CASE( testHashBoolVectorToBoolVectorCTX ) {
+    SHA256_CTX_mod ctx256;
+
+    std::vector<bool> preimage(SHA256_PREIMAGE_BYTES * 8);
+    libzerocash::convertBytesToVector(sha256_preimage, preimage);
+
+    std::vector<bool> expected(32*8);
+    libzerocash::convertBytesToVector(sha256_hash, expected);
+
+    // TODO: evaluate whether this should be a necessary precondition.
```

Comment:
What I'm worried about is this: `hashVector` uses the size of `actual` in the loop to copy the hash, so if you *don't* pre-allocate it before calling, you'll get an empty vector as the hash without any sort of error. That could obviously lead to security vulnerabilities.